admin/class-smaily-for-wp-admin.php

```diff
@@ -105,216 +111,270 @@ public function smaily_subscription_widget_init() {
-				);
+				$result = array_merge( $result, $this->remove_api_key() );
 				break;
-
 			case 'resetForm':
-				// Generate form contents.
-				$template = new Smaily_For_WP_Template( 'public/partials/smaily-for-wp-public-advanced.php' );
-
-				// Load configuration data.
-				$table_name = esc_sql( $wpdb->prefix . 'smaily_config' );
-				$data       = $wpdb->get_row( "SELECT * FROM `$table_name` LIMIT 1" );
-				$data->form = '';
-				$template->assign( (array) $data );
-
-				// Render template.
-				$result = array(
-					'error'   => false,
-					'message' => __( 'Newsletter subscription form reset to default.', 'smaily-for-wp' ),
-					'content' => $template->render(),
-				);
+				$result = array_merge( $result, $this->reset_form() );
 				break;
-
 			case 'save':
-				// Get parameters.
-				$isAdvanced = ( isset( $form_data['is_advanced'] ) && ! empty( $form_data['is_advanced'] ) ) ? '1' : '0';
-				$advanced   = ( isset( $form_data['advanced'] ) && is_array( $form_data['advanced'] ) ) ? $form_data['advanced'] : array();
-				$form       = ( isset( $advanced['form'] ) && is_string( $advanced['form'] ) ) ? $advanced['form'] : '';
-
-				// Generate new form (if empty).
-				if ( empty( $form ) ) {
-					$template = new Smaily_For_WP_Template( 'public/partials/smaily-for-wp-public-advanced.php' );
-
-					// Load configuration data.
-					$table_name = esc_sql( $wpdb->prefix . 'smaily_config' );
-					$data       = $wpdb->get_row( "SELECT * FROM `$table_name` LIMIT 1" );
-					$template->assign( (array) $data );
-
-					// Render template.
-					$form = $template->render();
-				}
-
-				// Update configuration.
-				$table_name = esc_sql( $wpdb->prefix . 'smaily_config' );
-				$wpdb->query(
-					$wpdb->prepare(
-						"
-					UPDATE `$table_name`
-					SET `form` = %s, `is_advanced` = %d
-					",
-						$form,
-						$isAdvanced
-					)
+				$result = array_merge( $result, $this->save( $form_data ) );
+				break;
+		}
+
+		if ( $refresh && $result['error'] === false ) {
+			$result['content'] = $this->generate_admin_form();
+		}
+
+		echo wp_json_encode( $result );
+		wp_die();
+	}
+
+	/**
+	 * Regenerate admin form using options stored in DB.
+	 *
+	 * @since  3.0.0
+	 * @access private
+	 * @return string HTML of admin form.
+	 */
+	private function generate_admin_form() {
+		// Generate form contents.
+		$template = new Smaily_For_WP_Template( 'admin/partials/smaily-for-wp-admin-form.php' );
+
+		// Load configuration data.
+		$api_credentials = get_option( 'smailyforwp_api_option' );
+		$form_options    = get_option( 'smailyforwp_form_option' );
+		$credentials_set = ! empty( $api_credentials['subdomain'] ) && ! empty( $api_credentials['username'] ) && ! empty( $api_credentials['password'] );
+
+		$template->assign(
+			array(
+				'form'                => $form_options['form'],
+				'api_credentials_set' => $credentials_set,
+			)
+		);
+
+		// Render template.
+		return $template->render();
+	}
```

Comment:
Make this function reusable, by:
- providing `$api_credentials_set` and `$form` as function arguments;
- returning `$template`.

This would allow you to get rid of loading configuration data as you would most likely is already loaded.

Same logic could be applied for `basic` and `advanced` form rendering.

Reply:
Added `generate_admin_template` and `generate_signup_template`